File: 程式設計：資料結構答案檔/C/103.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Monster {
    int Value;
    struct Monster* Next;
} Monster;

Monster* createMonster(int value) {
    Monster* newMonster = (Monster*)malloc(sizeof(Monster));
    if (newMonster == NULL) {
        exit(1); // Allocation failed
    }
    newMonster->Value = value;
    newMonster->Next = NULL;
    return newMonster;
}

void addMonster(Monster** head, int value) {
    Monster* newNode = createMonster(value);
    if (*head == NULL) {
        *head = newNode;
    } else {
        Monster* current = *head;
        while (current->Next != NULL) {
            current = current->Next;
        }
        current->Next = newNode;
    }
}

void createMonsters(Monster** head, const char* input) {
    char* token = strtok((char*)input, ",");
    while (token != NULL) {
        int monsterAttack = atoi(token);
        addMonster(head, monsterAttack);
        token = strtok(NULL, ",");
    }
}

int simulateBattle(Monster* head, int playerHP) {
    Monster* current = head;
    int level = 1;

    while (current != NULL) {
        playerHP -= current->Value;

        if (playerHP <= 0) {
            printf("game over! dead at level %d\n", level);
            return 0;
        }

        level++;
        current = current->Next;
    }

    printf("game clear! hp left %d\n", playerHP);
    return 1;
}

void freeMonsters(Monster* head) {
    while (head != NULL) {
        Monster* temp = head;
        head = head->Next;
        free(temp);
    }
}

int main() {
    char input[256];
    if (fgets(input, sizeof(input), stdin) == NULL) {
        printf("input error\n");
        return 1;
    }

    // Remove newline character if present
    input[strcspn(input, "\n")] = 0;

    Monster* head = NULL;
    createMonsters(&head, input);

    int playerHP = 100;
    simulateBattle(head, playerHP);

    freeMonsters(head);

    return 0;
}

File: 程式設計：資料結構答案檔/C#/103.cs
using System;

class Monster
{
    public int Value { get; set; }
    public Monster Next { get; set; }

    public Monster(int value)
    {
        Value = value;
        Next = null;
    }
}

class BattleSimulator
{
    private Monster head;

    public BattleSimulator()
    {
        head = null;
    }

    public void CreateMonsters(string input)
    {
        string[] values = input.Split(',');

        foreach (string value in values)
        {
            if (int.TryParse(value, out int monsterAttack))
            {
                AddMonster(monsterAttack);
            }
            else
            {
                Console.WriteLine("input error");
            }
        }
    }

    public void AddMonster(int value)
    {
        Monster newNode = new Monster(value);

        if (head == null)
        {
            head = newNode;
        }
        else
        {
            Monster current = head;
            while (current.Next != null)
            {
                current = current.Next;
            }
            current.Next = newNode;
        }
    }

    public bool SimulateBattle(int playerHP)
    {
        Monster current = head;
        int level = 1;

        while (current != null)
        {
            playerHP -= current.Value;

            if (playerHP <= 0)
            {
                Console.WriteLine(string.Format("game over! dead at level {0}", level));
                return false; 
            }

            level++;
            current = current.Next;
        }

        Console.WriteLine(string.Format("game clear! hp left {0}", playerHP));
        return true; 
    }
}


class Program
{
    static void Main()
    {
        string input = Console.ReadLine();

        BattleSimulator battle = new BattleSimulator();
        battle.CreateMonsters(input);

        int playerHP = 100;
        battle.SimulateBattle(playerHP);
    }

}
